runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3198,551 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+   generateReportFieldAccessOutlinedInstructions(node, mergePointLabel, dataSnippet, isWrite, cg);
+
+   outlinedSlowPath->swapInstructionListsWithCompilation();
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, mergePointLabel);
+
+   if (isResolved)
+      {
+      cg->stopUsingRegister(fieldClassReg);
+      }
+   cg->stopUsingRegister(fieldClassFlags);
+   }
+
+void
+J9::Z::TreeEvaluator::indirectReadWriteBarrierHelperForFieldWatch(TR::Node *node, J9Method *owningMethod, int32_t bcIndex, bool isWrite, TR::CodeGenerator *cg)
+   {
+   bool isResolved = !node->getSymbolReference()->isUnresolved();
+   J9JITWatchedInstanceFieldData instanceFieldDataBlock;
+
+   instanceFieldDataBlock.method = owningMethod;
+   instanceFieldDataBlock.location = bcIndex;
+
+   if (isResolved)
+      {
+      instanceFieldDataBlock.offset = static_cast<UDATA>(node->getSymbolReference()->getOffset() - TR::Compiler->om.objectHeaderSizeInBytes());
+      }
+   else
+      {
+      // If unresolved at compile time, we initialize the offset to -1. The call to generateFillInDataBlockSequenceForUnresolvedField will
+      // generate instructions to resolve and correctly populate this field at runtime.
+      instanceFieldDataBlock.offset = static_cast<UDATA>(-1);
+      }
+
+   TR::S390WritableDataSnippet *dataSnippet = static_cast<TR::S390WritableDataSnippet*> (cg->CreateConstant(node, &instanceFieldDataBlock, 1 << (int) (log2(sizeof(J9JITWatchedInstanceFieldData)-1)+1), true));
+
+   if (!isResolved)
+      {
+      // Resolve and populate dataSnippet fields.
+      generateFillInDataBlockSequenceForUnresolvedField(node, owningMethod, dataSnippet, isWrite, cg);
+      }
+
+   TR::Register *objReg = cg->evaluate(node->getFirstChild());
+   TR::Register *j9classReg = cg->allocateRegister();
+   TR::Register *j9classFlags = cg->allocateRegister();
+ 
+   // Load the class of the instance object into j9classReg.
+   TR::TreeEvaluator::genLoadForObjectHeadersMasked(cg, node, j9classReg, generateS390MemoryReference(objReg, static_cast<int32_t>(TR::Compiler->om.offsetOfObjectVftField()), cg), NULL);
+
+   TR::LabelSymbol *mergePointLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *fieldReportLabel = generateLabelSymbol(cg);
+
+   TR_ASSERT_FATAL(J9ClassHasWatchedFields >= std::numeric_limits<uint16_t>::min() && J9ClassHasWatchedFields <= std::numeric_limits<uint16_t>::max(), "expect value of J9ClassHasWatchedFields to be with in 16 bits");
+
+   // Test the bit in fieldClassReg with the J9ClassHasWatchedFields constant to make sure fieldwatch is enabled.
+   auto instr = generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, j9classFlags, generateS390MemoryReference(j9classReg, cg->comp()->fej9()->getOffsetOfClassFlags(), cg));
+   generateRIInstruction(cg, TR::InstOpCode::NILL, node, j9classFlags, J9ClassHasWatchedFields);
+   cg->setImplicitExceptionPoint(instr);
+   instr->setNeedsGCMap(0x0000FFFF);
```

Comment:
GC Point is not needed as it's redundant. Removed here: 5938c69

Reply:
We can't use a `TM` instruction specifically because the mask is a single byte and `J9ClassHasWatchedFields` is larger than one byte. 

We can however use a `TMLL` as the immediate value there is a 2-byte constant. It is also more correct to do `TMLL` than a `NILL`. I've done this here: c4a33cc

Note: We still have to do the Load though because `TMLL` is a Register-Immediate instruction so we need to load the flags into a register.

Reply:
Technically `J9ClassHasWatchedFields` is limited by the number of bits in the class flags which is 2 bytes as you say, however the the value should only have one active bit, so you should technically be able to use `TM` and avoid the load by calculating which bit you are testing and map that to either a `TM` on the first byte or the second byte of the class flags.

However this is such a small detail it's not worth it to implement at this point. I'm fine with the way things are.

Reply:
Yeah, that makes sense. I did actually try something similar to this yesterday. In the broader scope of things I don't think it gains anything in performance and unnecessarily adds more work at compile time.